design/cluster-api-provider-metal3/dhcpless_ipam.md

```diff
@@ -0,0 +1,397 @@
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: data-template
+      hostSelector:
+        matchLabels:
+          cluster-role: control-plane
+
+```
+
+In this flow we have `preprovisioning-pool` which is not associated with any cluster, this is used to provide an IPAddress during
+the pre-provisioning phase as described above.  To reduce the required size of the pool, the IPClaim will be deleted after the
+preprovisioning phase is completed, e.g the BMH resource becomes available.
+
+In the provisioning phase another pool, associated with a cluster is used to template networkData as in the existing process.
+
+#### Assumptions and Open Questions
+
+TODO
+
+### Inspection on initial registration
+
+On initial registration of a host, inspection is triggered immediately but this process cannot complete without preprovisioning network configuration in a DHCP-less environment (because the IPA ramdisk can't connect back to the Ironic API).
+
+To resolve this we can add a new BareMetalHost API `PreprovisioningNetworkDataRequired` which defaults to false, but when set to true will describe that the host cannot move from Registering -> Inspecting until `preprovisioningNetworkDataName` has been set.
+
+An alternative could be to require that the BareMetalHost resources are created with the existing [paused annotation](https://github.com/metal3-io/baremetal-operator/blob/main/docs/api.md#pausing-reconciliation), set to a pre-determined value (e.g `metal3.io/preprovisioning`) which can then be removed by the new controller after `preprovisioningNetworkDataName` has been set, then inspection will be able to succeed.
+
+### Implementation Details/Notes/Constraints
+
+#### IP Reuse
+
+A related issue has been previously addressed via the [IP Reuse](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/ip_reuse.md) functionality - this means we can couple IPClaims to the BareMetalHost resources which will enable consistent IP allocations for pre-provisioning and subsequent provisioning operations (provided the same IPPool is used for both steps)
+
+### Risks and Mitigations
+
+- TODO
+
+### Work Items
+
+TODO
+
+### Dependencies
+
+#### Firstboot agent support
+
+An agent in the IPA ramdisk image is required to consume the network data provided via the processes outlined above.
+
+The Ironic DHCP-less documentation describes using glean (a minimal python-based cloud-init alternative), but we don't
+currently have any community-supported IPA ramdisk image containing this tool.
+
+There are several other options such as cloud-init, or even custom scripts/tooling which may be coupled to the OS, so we
+do not define a specific solution as part of this proposal.
+
+#### Potential config-drive conflict on redeployment
+
```

Comment:
@Rozzii can you expand on `Ironic driver support for any of these approaches is not complete yet` please?  My testing related to https://github.com/metal3-io/baremetal-operator/pull/1380 indicates that packing the node network_data into the IPA virtualmedia image works fine, other than the config-2 redeploy-conflict issue discussed elsewhere.

Reply:
There is a situation where the IPA is a pre-built ISO for example with redfish-virtualmedia, so the combination of the redfish-virtualmedia + pre-built ISO + DHCPless environment would cause a challenge because as far as I know there are no functionality in redfish or any other Ironic driver that would take the json data from the network_data endpoint of an Ironic Node and mount it as a second block device with the config-2 or preferably with ir-vfd-dev or a custom  (to avoid config-2 collision) disk label.

Reply:
If I remember correcly if the IPA is supplied as kernel and initramfs combo then Ironic can convert it to an ISO with xorriso and it can inject the config data to the ISO. So that is why I am saying that the support is not complete.

Reply:
Thanks @Rozzii - if I am understanding correctly you're saying this approach will not work in some configurations, and specifically it won't work with `format: live-iso` since in that case Ironic doesn't build the virtualmedia image, so it can't inject any `network_data`

Reply:
I tried to capture this in the Drawbacks section, please let me know if I missed anything, thanks!